### ISSUE TITLE
jq for advanced transforms

### ISSUE THREAD

**Issue by broskees** [josdejong/jsoneditor#857]
JMESpath vs JQ
-
JMESpath is super easy to understand and I like that.
However, it has some serious limitations. Namely:
- No simple if/else
- The `to_number` function doesn't support boolean values
- can't reference parents when doing iteration
- Key name can't come from an expression
- No basic math operations, add, multiply, divide, mod, etc.
- There's a join, but no split
- No array indexing based on expression
- No ability to group_by an expression.
- No ability to get the index of an element in a list

I think this really prevents JMESpath from being as general purpose as [jq](https://stedolan.github.io/jq/manual/).

I thought that perhaps, by using the [jq-web](https://github.com/fiatjaf/jq-web) library, we could extend the transform feature to allow for jq queries also.

UI wise, I think that we could just add tabs for both JMESpath & jq at the top of the transform overlay.

Considering the limitations I listed above, I think that its fairly easy to say this would solve a lot of use cases where [jsoneditor](https://github.com/josdejong/jsoneditor) wouldn't have been powerful enough to solve the issue at hand.

Interested in hearing the communities' thoughts.

**Comment by josdejong** [comment]
Thanks for starting this discussion! I'm open to discuss alternative solutions.

I regularly encounter limitations with JMESPath, for example last week I wanted to get all distinct values in a list with strings which is not possible.

If I remember well, when choosing a query language I think I did not selected jq-web because I thought it was too large to embed (like 2 MB). Looking closer I _think_ that I was wrong, there there is a bundle [jq.wasm.min.js]( https://unpkg.com/browse/jq-web@0.5.1/) which is "only" 63.3 kB which may be acceptable to bundle in JSONEditor. jq is much more mainstream and actively maintained, so that can be a good choice. JMESPath on the other hand seems to be stale.

I recently saw another query language which can be interesting too: https://npm.taobao.org/package/groq-js. Though this is a work in progress, may be too early.

We can basically go in two directions:

1. Allow multiple query languages alongside each other (JMESPath, jq, Groq, etc.)
2. Replace the current JMESPath with an alternative solution.

I think we should not go for multiple query languages alongside each other, that will be only confusing for the user. It may be possible though to create an API so you can hook up your own favorite language to replace the built-in one. We should only go that route when it's really easy to replace implementations, else I don't think it's worth it.

**Comment by broskees** [comment]
I agree in regards to replacing JMESpath with jq as being a better solution. The only worry I have is that I'm sure many users of jsoneditor are quite used to JMESpath by now. That being said, I haven't found much at all as far as limitations with jq, so maybe the change will be welcome. API would be a nice work around if it's easy enough to work with.

In regards to the jq-web library. I'm using the jq.wasm.min.js library in a project now via npm. It's working well enough for my purposes, unfortunately though, it requires jq.wasm.wasm. This setup also requires using the library asynchronously, which has proven to be a bit annoying, although I'm pretty comfortable with that now. 

Wasn't aware of GROQ at all. After briefly looking at the syntax it looks awesome! Seems to solve a lot of the verbosity that jq can end up being with large queries too. However groc-js isn't even a stable release yet, so I'd wait on that one, though I'd be interested in finding out what the filesize footprint is for that.

**Comment by broskees** [comment]
Just thought of this. An alternative solution considering the footprint. What about just adding the API? That way, we don't have to natively bundle the jq-web library, but it can be added on when it's needed.

**Comment by josdejong** [comment]
Ow, too bad that `jq` is so large after all :( 

Let's indeed first explore if it is an option to create an API for this :+1:

**Comment by broskees** [comment]
Any ideas on how we'd structure it?

**Comment by josdejong** [comment]
I've had a look at the code, and I think it we can create a beautiful and simple API for this!

The API need two things:

1. A function to build a query from the transform wizard form, something like:
    ```js
    const query = createQuery({ 
      filterBy: { field, relation, value },
      sortBy: { field, direction },
      pickFields: [ field1, field2, ... ]
    })
    ```

2. A function to actually transform JSON for a given query (execute a query):
    ```js
    const newJSON = executeQuery(oldJSON, query)
    ```

**Comment by broskees** [comment]
I can start working on a PR. Another question though, how will the createQuery and executeQuery functions know what querying languages to use?

**Comment by josdejong** [comment]
Thanks for your offer! 

If you provide your own implementation of `createQuery` and `executeQuery`, this implementation will import and use your query engine of choice (like `jq`), so I think JSONEditor does not need to be aware of that. Or am I overlooking something here?

**Comment by broskees** [comment]
Maybe i'm not understanding what the executeQuery function is supposed to do.

If it's supposed to literally execute a jq query (or any other querying language for that matter), it would have to at least need the processor passed to it, right?

**Comment by broskees** [comment]
Oh okay, that makes complete sense. Where does this proof of concept need to be improved to be ready for a PR?

**Comment by josdejong** [comment]
There are a couple of details that need to be worked out, like there is still some reliance on JMESPath specific `@`, and we should be able to customize the explanatory text in the query modal.

It may be easiest if I work out the PR of #871. It's holiday period so I have some time for this :)

**Comment by josdejong** [comment]
@broskees I've worked out docs, examples, and details for custom query language support in #871, this is now ready to be merged. It would be great if you can give it a try and see if it makes sense and works as you would expect.

I have to say, I like the JavaScript example [23_custom_query_language.html](https://github.com/josdejong/jsoneditor/blob/d12a5e7cf29f42ec8ec3da36bd8fa8f66fdfd4ea/examples/23_custom_query_language.html) a _lot_ over the built-in JMESPath query language. Being able to write plain JavaScript is much more powerful and feels more natural to me. There is one drawback though: it can easily give security issues if you allow executing arbitrary text as JavaScript. I have to think about that.

**Comment by broskees** [comment]
Thanks! I'll try incorporating this into the relevant project I'm working on and I'll lyk how it works!

**Comment by josdejong** [comment]
Customizing the query language is now available in `v8.5.0`.
